docs/source/components.rst

```diff
@@ -0,0 +1,332 @@
+be in the future) and cross field validation (e.g. if field A is set field B
+must be unset and vice versa, or field ``first_number`` and field
+``second_number`` added up must be smaller than 100).
+
+This component works only with the universal data structure mentioned before and
+should not fetch additional data itself. It should however be possible for the
+validation component to be influenced by the environment (a user's permission,
+current time, etc.) and change its logic based on that.
+
+Additionally a validator should also be allowed to change values in a structural
+way to make coercing of values possible. Examples are casting a string to an
+integer or normalizing a unicode string.
+
+
+How it is currently done
+------------------------
+
+Currently there are many ways to specify field validation in Django. The easiest
+is to pass validator functions to the field via the ``validators`` argument:
+
+.. code-block:: python
+
+    title = CharField(validators=[
+        validate_illegal_characters,
+        validate_banned_words,
+    ])
+
+
+Where the validation functions just take a value and raise a ``ValidationError``
+if applicable.
+
+Custom fields can also override ``clean()``. This method can also change the
+value that is validated.
+
+.. code-block:: python
+
+    class TitleField(CharField):
+        def clean(self, value):
+            value = super().clean(value)
+            if not value.startswith('Title'):
+                # all titles must start with "Title"
+                value = 'Title ' + value
+            if len(value.split()) > 5:
+                raise ValidationError(
+                    'title must not contain more than 5 words'
+                )
+            return value
+
+
+It is also possible to define field validators on a form by adding
+``clean_<field_name>()`` methods to it:
+
+.. code-block:: python
+
+    class BlogPostForm(Form):
+        title = CharField()
+        content = CharField()
+
+        def clean_title(self):
+            title = self.cleaned_data['title']
+            if 'buzzword' in title.lower():
+                raise ValidationError('invalid word')
+            return title
+
+
+Just like a field's ``clean()`` method this method can also change the value.
+
+Cross field validation is made possible by overriding ``Form.clean()``:
+
+.. code-block:: python
+
+    class NumbersForm(Form):
+        first_number = IntegerField()
+        second_number = IntegerField()
+
+        def clean(self):
+            data = super().clean()
+            if data['first_number'] + data['second_number'] > 100:
+                raise ValidationError(
+                    'sum of numbers must be smaller than 100'
+                )
+            return data
+
+
+How it could be done
+--------------------
+
+There are several libraries that explicitly deal with validation in Django.
+
+
+Rendering
+=========
+
+Whenever the universal data structure or a subset thereof should be send or
```

Comment:
Serializer Renderers receive the underlying data structure and the current state of the serializer data, presenting them to the outside world. This could be as an HTML form, a JSON object or something completely different.